---
title: "Final_Project"
author: "Connor Gagen, Julia Keefe, Paloma Peters, Anna Mellizo Kroll, Anna Broerman"
date: "2025-03-17"
output: html_document
editor_options: 
  chunk_output_type: console
---
## Goal to analyze time course data
I am looking at mouse MM10 with data where dox added to stem cell mouse then RNA was collected in 3 replicates and sequenced

RAW sequecncefiles are here: 
File path/ 

These fastqs were run through NF_CORE RNAS seq version X
### Multiqc html is here: 
file path/ 

### Outpdir NF_CORE is here: 
filepath/ 

counts_matrix <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/data/salmon.merged.gene_counts.tsv", header=TRUE, row.names=1)

TPM <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/data/salmon.merged.gene_tpm.tsv", header=TRUE, row.names=1)

### Other output NF_CORE Testing is here

# DESEQ was run with the following unique_sig_genes, counts_integer, g2s, res_df, filtered_res_df, and deseq_samples results:
load("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/lessons/06_Differential_expression_analyses/project/project_results/counts_DESEQ_results.rdata")

# TPM, average_and_stddev_values, TPM_filtered results:
load("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/lessons/06_Differential_expression_analyses/project/project_results/TPM_results.Rdata")

# We decided to look at the top 5 genes with the largest log2fold change and smallest standard deviation to see how they differentiate:
Gm45234 *min
Gm13694 *max

H19 *another gene that could be good*

# creating a heat map to visualize sig gene expression
* note we use row centering and not 0 normalization. 

## load data for heatmap


```

# (1) Cluster samples
First step is calculating distance between all samples. Let's start with our
TPM data as these are the values typically used for clustering and heatmaps. 

# Note use of transmute (t) 
# Note log TPM values being used
```{r distance calculation}

# Log-transform the TPM values. 
# This helps deflate spurious and low values - common practice for clustering/heatmaps
log_tpm_matrix <- log2(TPM_filtered + 1)

# Now let's calculate how closely related each sample (time point) is to each other using DIST function.
# We make a vector the length of genes in TPM_filtered. Then use correlation (DIST function) to find similarity.
# For other methods check out dist menu 
?dist

# Let's start with euclidean distance / similarity based on log2 TPM values of genes in TPM_filtered.
distance_matrix <- dist(t(log_tpm_matrix), method = "euclidean")

# Now we use the function : HCLUST ! 
# This performs the hierarchical clustering of each samples relationship
?hclust
# Other forms of clustering as well see menu above for now ward or complete
# Downside of ward is it assumes a circle relationship where as complete is linear - like we have with time.
hc_complete <- hclust(distance_matrix, method = "complete")

# Now we can use plot() to plot the distance calucalated in hclust!
plot(hc_complete, labels = colnames(log_tpm_matrix), main = "Global TPM Filtered Dendrogram ", sub = "sample relationship by time and replicate after dox treatment")
# Cool we see that samples see to return closer to 0 after 96 hours of dox treatment
# Also that the replicates are more similar than time points - a very good / expected thing 

# Cut tree to limit the clusters 
clusters <- cutree(hc_complete, k = 5)
# Print out which sample is in which cluster
print(clusters)

# Results: WT-
```