---
title: "06_knitting_ATAC_analyses"
author: "JR"
date: "2024-11-30"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/util/00_our_first_function_lesson.R")
library(GenomicRanges)
library(IRanges)
library(dplyr)
library(tidyr)
library(tibble)
library(readr)
library(ggplot2)
library(purrr)
library(magrittr)
library(pheatmap)
library(textshape)
library(Rcpp)
library(DESeq2)
library(stringr)
```
#### Objective: Analyse the ATACSeq data for the human data

### Use DESEQ2 to find out if peaks are changing in dox and non-dox conditions
```{r create input files and run DESEQ2 on feature counts of consensus peaks from NF_CORE for human data, dependson="previous_chunk"}

# Consensus peaks
broad_consensus_peaks_human <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/DATA/human_atacseq/consensus_peaks.mLb.clN.annotatePeaks.txt",
                             sep = "\t", header = TRUE)

# Consensus peak counts
broad_consensus_counts_human <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/DATA/human_atacseq/consensus_peaks.mLb.clN.featureCounts.txt",
                             sep = "\t", header = TRUE)

# Creating sample sheet "atac_samples" from file names
broad_consensus_counts_human <- broad_consensus_counts_human %>%
  dplyr::select(-c(Geneid, Chr, Start, End, Strand, Length))
colnames(broad_consensus_counts_human) <- gsub("_REP1",
                                         "_R1",
                                         gsub(
                                           "\\.mLb\\.clN\\.sorted\\.bam",
                                           "",
                                           colnames(broad_consensus_counts_human)
                                         ))
count_columns_human <- colnames(broad_consensus_counts_human)

atac_samples_human <- data.frame(
  sample = count_columns_human,
  condition = ifelse(
    grepl("_0h_", count_columns_human),
    "non-dox",
    ifelse(grepl("non-dox", count_columns_human), "non-dox", "dox")
  ),
  timepoint_minutes = as.numeric(sub(".*_(\\d+)_R1.*", "\\1", count_columns_human))
)

# Factor condition for DESEQ2
atac_samples_human <- atac_samples_human %>%
  mutate(condition = factor(condition, levels = c("non-dox", "dox")))

# Matrix for Deseq2
atac_dds_condition_human <- suppressWarnings(DESeqDataSetFromMatrix(countData = broad_consensus_counts_human, 
                                   colData = atac_samples_human, 
                                   design = ~ condition))
# Run DESeq2 condition model
atac_dds_condition_human <- suppressWarnings(DESeq(atac_dds_condition_human))

```

### Extracting and analyzing DESEQ2 results
```{r extracting DESEQ2 results in human data, dependson="previous_chunk"}

# Extract DESeq2 results
atac_lfc_condition_human <- results(atac_dds_condition_human) %>%
  as.data.frame() %>%
  rownames_to_column("interval_id")

# Merge with broad_consensus_peaks info
colnames(broad_consensus_peaks_human)[1] <- "interval_id"

atac_lfc_condition_human <- merge(atac_lfc_condition_human, 
                  broad_consensus_peaks_human %>%
                    dplyr::select(interval_id, Gene.Name, Nearest.PromoterID, 
                                  Distance.to.TSS, Chr, Start, End),
                  by = "interval_id")

```

### Analyzing results of DESEQ on ATACseq peak counts in human data
```{r analysis of DESEQ2 on ATAC peak counts in human data, dependson="previous_chunk"}
hist(atac_lfc_condition_human$padj)
```