---
title: "Final CGL Project"
author: "Paloma Peters, Julia Keefe, Connor Gagen, Anna Mellizo Kroll, Anna Broerman"
date: "2024-04-27"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/util/00_our_first_function_lesson.R")
library(GenomicRanges)
library(IRanges)
library(dplyr)
library(tidyr)
library(tibble)
library(readr)
library(ggplot2)
library(purrr)
library(magrittr)
library(pheatmap)
library(textshape)
library(Rcpp)
library(DESeq2)
library(stringr)
library(knitr)
```

## Overall Objective: To compare RNASeq data to ATACSeq data and determine if dox has an effect on human and mouse genomic expression

### Objective 1: RNAseq analysis to find genes changing significantly across time from dox exposure in mice

#### Importing counts values
```{r import counts from Salmon}
# This loads in the counts from the NF core run
counts_matrix <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/data/salmon.merged.gene_counts.tsv", header=TRUE, row.names=1)

# This will prepare the counts for DESEQ by creating g2s
g2s <- data.frame(
  gene_id = rownames(counts_matrix),
  gene_name = counts_matrix[, 1]
)
# This removes the gene name from the counts matrix such that it is only numeric values
counts_matrix <- counts_matrix[ , -1]
# Round the values within the counts matrix. 
counts_integer <- round(counts_matrix)
```

#### Creating sample sheet for DESeq from counts file
```{r creating DESeq sample sheet}
# This creates the data frame with the same columns as the counts_matrix
deseq_samples <- data.frame(
  sample_id = colnames(counts_matrix))
# Splits the sample_id to create a columns for time point and replicate columns
split_values <- strsplit(deseq_samples$sample_id, "_")
# This extracts the time point and the replicate info
time_values <- sapply(split_values, function(x) x[[2]])
replicate_values <- sapply(split_values, function(x) x[[3]])
# This will place the time point and replicate into the data frame.
deseq_samples$time_point <- time_values
deseq_samples$replicate <- replicate_values
# This converts the time point and replicate data into a factor.
deseq_samples$time_point <- factor(deseq_samples$time_point)
deseq_samples$replicate <- factor(deseq_samples$replicate)
```

#### Running DESEQ2
```{r running deseq2}
# Test to see if the sample sheet matches the corresponding deseq_samples data frame
stopifnot(all(colnames(counts_integer) == rownames(deseq_samples$sample_id)))
# Sets up the dds
dds_time_point <- DESeqDataSetFromMatrix(countData = counts_integer,
                              colData = deseq_samples,
                              design = ~ time_point)
# Runs the DESeqq
dds_time_point <- DESeq(dds_time_point)
```

#### Extracting results for each time point from DESEQ dds
```{r compile DESEQ2 results}
# Retrieves the names corresponding to the different time points 
result_names <- resultsNames(dds_time_point)
# Removes the first entry from deseq which is the baseline. 
results_names <- result_names[-1]

# Setting up data frame 
res_df <- data.frame("gene_id" = character(), 
                     "baseMean" = numeric(), 
                     "log2FoldChange" = numeric(), 
                     "lfcSE" = numeric(),
                     "stat" = numeric(),
                     "pvalue" = numeric(),
                     "padj" = numeric(),
                     "gene_name" = character(),
                     "result_name" = character())

# Run for loop to retreive results from all time comparisons in dds
for(i in 1:length(results_names)) {
  results_name <- results_names[i]
  res <- results(dds_time_point, name = results_name)
  tmp_res_df <- res %>% 
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    merge(g2s) %>%
    mutate(result_name = results_name)
  res_df <- bind_rows(res_df, tmp_res_df)
  
}
```

####  Filtering to significant results P <0.05 and LFC abs(1)
```{r filtering DESEQ results to LFC > abs(1)& P <0.05}
# Filters for padj and logfold
filtered_res_df <- res_df %>%
  filter(padj < 0.05, abs(log2FoldChange) > 1)
# Gives unique genes
unique_sig_genes <- unique(filtered_res_df$gene_id)
```
These unique_sig_genes found 657 genes that met the criteria for significance and were saved into our results folder so all 657 genes could be viewed in a .csv

### Objective 2: RNASeq analysis to find genes changing significantly across time from dox exposure in human

#### Importing counts values
```{r import counts from Salmon human}
# This loads in the counts from the NF core run
counts_matrix_human <- read.table("//scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/lessons/06_Differential_expression_analyses/project/Project2_files/salmon.merged.gene_counts.tsv", header=TRUE, row.names=1)

# This will prepare the counts for DESEQ by creating g2s
g2s_human <- data.frame(
  gene_id_human = rownames(counts_matrix_human),
  gene_name_human = counts_matrix_human[, 1]
)
# This removes the gene name from the counts matrix such that it is only numeric values
counts_matrix_human <- counts_matrix_human[ , -1]
# Round the values within the counts matrix. 
counts_integer_human <- round(counts_matrix_human)
```

#### Creating sample sheet for DESeq from counts file
```{r creating DESeq sample sheet human}
# This creates the data frame with the same columns as the counts_matrix
deseq_samples_human <- data.frame(
  sample_id_human = colnames(counts_matrix_human))
# Splits the sample_id to create a columns for time point and replicate columns
split_values_human <- strsplit(deseq_samples_human$sample_id_human, "_")
# This extracts the time point and the replicate info
time_values_human <- sapply(split_values_human, function(x) x[[2]])
replicate_values_human <- sapply(split_values_human, function(x) x[[3]])
# This will place the time point and replicate into the data frame.
deseq_samples_human$time_point_human <- time_values_human
deseq_samples_human$replicate_human <- replicate_values_human
# This converts the time point and replicate data into a factor.
deseq_samples_human$time_point_human <- factor(deseq_samples_human$time_point_human)
deseq_samples_human$replicate_human <- factor(deseq_samples_human$replicate_human)
```

#### Running DESEQ2
```{r running deseq2 human}
# Test to see if the sample sheet matches the corresponding deseq_samples data frame
stopifnot(all(colnames(counts_integer_human) == deseq_samples_human$sample_id_human))
rownames(deseq_samples_human) <- deseq_samples_human$sample_id_human
# Sets up the dds
dds_time_point_human <- DESeqDataSetFromMatrix(countData = counts_integer_human,
                              colData = deseq_samples_human,
                              design = ~ time_point_human)
# Runs the DESeqq
dds_time_point_human <- DESeq(dds_time_point_human)
```

#### Extracting results for each time point from DESEQ dds
```{r compile DESEQ2 results human}
# Retrieves the names corresponding to the different time points and Removes the first entry from deseq 
results_names_human <- resultsNames(dds_time_point_human)[-1]

# Setting up data frame 
res_df_human <- data.frame("gene_id_human" = character(), 
                     "baseMean" = numeric(), 
                     "log2FoldChange" = numeric(), 
                     "lfcSE" = numeric(),
                     "stat" = numeric(),
                     "pvalue" = numeric(),
                     "padj" = numeric(),
                     "gene_name_human" = character(),
                     "result_name_human" = character())

# Run for loop to retreive results from all time comparisons in dds
for(i in 1:length(results_names_human)) {
  results_name_human <- results_names_human[i]
  res_human <- results(dds_time_point_human, name = results_name_human)
  tmp_res_df_human <- res_human %>% 
    as.data.frame() %>%
    rownames_to_column("gene_id_human") %>%
    merge(g2s_human) %>%
    mutate(result_name_human = results_name_human)
  res_df_human <- bind_rows(res_df_human, tmp_res_df_human)
  
}
```

####  Filtering to significant results P <0.05 and LFC abs(1)
```{r filtering DESEQ results to LFC > abs(1)& P <0.05 human}
# Filters for padj and logfold
filtered_res_df_human <- res_df_human %>%
  filter(padj < 0.05, abs(log2FoldChange) > 1)
# Gives unique genes
unique_sig_genes_human <- unique(filtered_res_df_human$gene_id_human)

```
These unique_sig_genes found 7338 genes that met the criteria for significance and were saved into our results folder so all 7338 genes could be viewed in a .csv. A possible reason for there being lots more significant genes in the human data in comparison to the mouse data is because there was data from more time points. We still need to look more into these results, we ran out of time to double check everything.

### Objective 3: Analyse the ATACSeq data for the mouse data

#### Use DESeq2 to find out if peaks are changing in dox and non-dox conditions
```{r create input files and run DESEQ2 on feature counts of consensus peaks from NF_CORE, dependson="previous_chunk"}
# Consensus peaks
broad_consensus_peaks <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/08_ATACseq_pipeline/00_run_pipeline/00_pipeline_run/pipeline_run/bwa/merged_library/macs2/broad_peak/consensus/consensus_peaks.mLb.clN.annotatePeaks.txt",
                             sep = "\t", header = TRUE)

# Consensus peak counts
broad_consensus_counts <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/08_ATACseq_pipeline/00_run_pipeline/00_pipeline_run/pipeline_run/bwa/merged_library/macs2/broad_peak/consensus/consensus_peaks.mLb.clN.featureCounts.txt",
                             sep = "\t", header = TRUE)

# Creating sample sheet "atac_samples" from file names
rownames(broad_consensus_counts) <- broad_consensus_counts$Geneid

broad_consensus_counts <- broad_consensus_counts %>%
  dplyr::select(-c(Geneid, Chr, Start, End, Strand, Length))
colnames(broad_consensus_counts) <- gsub("_REP1",
                                         "_R1",
                                         gsub(
                                           "\\.mLb\\.clN\\.sorted\\.bam",
                                           "",
                                           colnames(broad_consensus_counts)
                                         ))
count_columns <- colnames(broad_consensus_counts)

atac_samples <- data.frame(
  sample = count_columns,
  condition = ifelse(
    grepl("_0_", count_columns),
    "non-dox",
    ifelse(grepl("non-dox", count_columns), "non-dox", "dox")
  ),
  timepoint_minutes = as.numeric(sub(".*_(\\d+)_R1.*", "\\1", count_columns))
)

# Factor condition for DESEQ2
atac_samples <- atac_samples %>%
  mutate(condition = factor(condition, levels = c("non-dox", "dox")))

# Matrix for Deseq2
atac_dds_condition <- suppressWarnings(DESeqDataSetFromMatrix(countData = broad_consensus_counts, 
                                   colData = atac_samples, 
                                   design = ~ condition))
# Run DESeq2 condition model
atac_dds_condition <- suppressWarnings(DESeq(atac_dds_condition))
```

#### Analyzing DESEQ2 results
```{r extracting DESEQ2 results, dependson="previous_chunk"}

# Extract DESeq2 results
atac_lfc_condition <- results(atac_dds_condition) %>%
  as.data.frame() %>%
  rownames_to_column("interval_id")

# Merge with broad_consensus_peaks info
colnames(broad_consensus_peaks)[1] <- "interval_id"

atac_lfc_condition <- merge(atac_lfc_condition, 
                  broad_consensus_peaks %>%
                    dplyr::select(interval_id, Gene.Name, Nearest.PromoterID, 
                                  Distance.to.TSS, Chr, Start, End),
                  by = "interval_id")

```

#### Analyzing results of DESeq on ATACseq peak counts
```{r analysis of DESEQ2 on ATAC peak counts, dependson="previous_chunk"}
hist(atac_lfc_condition$padj)
```
This histogram shows all the padj values around 1, which means none of them are significant.

### Objective 4: Analyse the ATACSeq data for the human data

#### Use DESEQ2 to find out if peaks are changing in dox and non-dox conditions
```{r create input files and run DESEQ2 on feature counts of consensus peaks from NF_CORE for human data human, dependson="previous_chunk"}
# Consensus peaks
broad_consensus_peaks_human <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/DATA/human_atacseq/consensus_peaks.mLb.clN.annotatePeaks.txt",
                             sep = "\t", header = TRUE)

# Consensus peak counts
broad_consensus_counts_human <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/DATA/human_atacseq/consensus_peaks.mLb.clN.featureCounts.txt",
                             sep = "\t", header = TRUE)

# Creating sample sheet "atac_samples" from file names
broad_consensus_counts_human <- broad_consensus_counts_human %>%
  dplyr::select(-c(Geneid, Chr, Start, End, Strand, Length))
colnames(broad_consensus_counts_human) <- gsub("_REP1",
                                         "_R1",
                                         gsub(
                                           "\\.mLb\\.clN\\.sorted\\.bam",
                                           "",
                                           colnames(broad_consensus_counts_human)
                                         ))
count_columns_human <- colnames(broad_consensus_counts_human)

atac_samples_human <- data.frame(
  sample = count_columns_human,
  condition = ifelse(
    grepl("_0h_", count_columns_human),
    "non-dox",
    ifelse(grepl("non-dox", count_columns_human), "non-dox", "dox")
  ),
  timepoint_minutes = as.numeric(sub(".*_(\\d+)_R1.*", "\\1", count_columns_human))
)

# Factor condition for DESEQ2
atac_samples_human <- atac_samples_human %>%
  mutate(condition = factor(condition, levels = c("non-dox", "dox")))

# Matrix for Deseq2
atac_dds_condition_human <- suppressWarnings(DESeqDataSetFromMatrix(countData = broad_consensus_counts_human, 
                                   colData = atac_samples_human, 
                                   design = ~ condition))
# Run DESeq2 condition model
atac_dds_condition_human <- suppressWarnings(DESeq(atac_dds_condition_human))
```

#### Extracting and analyzing DESEQ2 results
```{r extracting DESEQ2 results in human data human, dependson="previous_chunk"}

# Extract DESeq2 results
atac_lfc_condition_human <- results(atac_dds_condition_human) %>%
  as.data.frame() %>%
  rownames_to_column("interval_id")

# Merge with broad_consensus_peaks info
colnames(broad_consensus_peaks_human)[1] <- "interval_id"

atac_lfc_condition_human <- merge(atac_lfc_condition_human, 
                  broad_consensus_peaks_human %>%
                    dplyr::select(interval_id, Gene.Name, Nearest.PromoterID, 
                                  Distance.to.TSS, Chr, Start, End),
                  by = "interval_id")

```

#### Analyzing results of DESEQ on ATACseq peak counts in human data
```{r analysis of DESEQ2 on ATAC peak counts in human data, dependson="previous_chunk"}
table(atac_samples_human$condition)
```
Our DESeq2 could not extract any meaningful results because of the difference in the dox vs. non-dox sample sizes. This meant when running:

atac_lfc_condition_human <- results(atac_dds_condition_human) %>%
  as.data.frame() %>%
  rownames_to_column("interval_id")
  
it generated an empty atac_lfc_condition_human file so a histogram of padj couldn't be generated.

### Objective 5: To look at the 27 genes that overlap between human and mouse data from the RNASeq analysis

#### We got these 27 genes from another group in our class. These genes are the overlaps between the significant genes from the RNASeq data for both human and mouse
```{r 27 sig genes image, echo=FALSE, fig.cap="The 27 Overlapping Genes"}
include_graphics("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/lessons/06_Differential_expression_analyses/project/Figures/27genes.png")
```

#### Of the 27 genes 8 were found on IGV. We looked at the common peaks ATACSeq data for those 8 genes and compared the IGV results for both the human and mouse data
```{r IGV image, echo=FALSE, fig.cap="IGV results"}
include_graphics("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/lessons/06_Differential_expression_analyses/project/Figures/table1.png")
include_graphics("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/lessons/06_Differential_expression_analyses/project/Figures/table2.png")
include_graphics("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/lessons/06_Differential_expression_analyses/project/Figures/table3.png")
include_graphics("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/lessons/06_Differential_expression_analyses/project/Figures/table4.png")

```

#### Conclusions:
As seen in the ATACSeq graphs none of the p-values are significant, they're all about equal to 1. This means that there was basically no significant difference in chromatin accessibility between dox and non-dox. Also, if you look at the IGV results of the 8 different genes we looked at none of them had common peaks between the human and mouse data. This shows the differences in chromatin accessibility are not conserved between human and mouse. 