---
title: "Project Part 2"
author: "Connor Gagen, Julia Keefe, Paloma Peters, Anna Mellizo Kroll, Anna Broerman"
date: "2024-11-30"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/util/00_our_first_function_lesson.R")
library(GenomicRanges)
library(IRanges)
library(dplyr)
library(tidyr)
library(tibble)
library(readr)
library(ggplot2)
library(purrr)
library(magrittr)
library(pheatmap)
library(textshape)
library(Rcpp)
library(DESeq2)
library(stringr)
```

#### Objective: To determine if chromatin accessibility changes due to dox exposure.
We have performed a time course series of experiements measuring chromating accessibility
(ATACseq peaks)upon exposure to dox. These are 0, 30, 60, 90, 120, 150 minutes.
Each time point has a replicate. The fastq sequencing files were processed by the
NF_CORE ATACseq pipeline v- 2.12 (https://nf-co.re/atacseq/2.1.2/).

We will be using the MACS2 output files from the NF_CORE Pipeline For example:
-broad.peak
-consensus.broad.peak
-consensus.feature.counts

#### Approach:

##### (a) Loading in ATACseq peak files with custom function import_peaks (list of GRanges output)

```{r loading in peak files to list of GRanges}

# establishing peak path to the dir with MACS2 output peak files from NF_CORE ATACseq pipeline

peak_path <- "/scratch/Shares/rinnclass/MASTER_CLASS/lessons/08_ATACseq_pipeline/00_run_pipeline/00_pipeline_run/pipeline_run/bwa/merged_library/macs2/broad_peak"

# creating a file list also needed for import_peaks function to get sample name associated with file
fl <- list.files(peak_path, full.names = TRUE, pattern = ".broadPeak")

# running import_peaks
my_peaks <- import_peaks(consensus_file_path = peak_path)

```

##### (b) finding number of peaks common in all samples using find_common_peaks custom funciton
 
```{r finding common peaks in all samples, dependson="previous_chunk"}
# run find_common_peaks function 
common_peaks <-  suppressWarnings(find_common_peaks(my_peaks))

```


##### (c) Peaks that are unique to dox and non-dox conditions
Using find_common_overlaps peaks that are specific to dox or non-dox
will be identified.

Non-dox samples
```{r non-dox atac peaks, dependson="previous_chunk"}

# common peaks in non-dox (0 time point)
non_dox_samples <- my_peaks[c("KO_control_0", "WT_control_0")]
non_dox_common_peaks <- suppressWarnings(find_common_peaks(non_dox_samples))
```

Dox samples
```{r dox samples, dependson="previous_chunk"}
# common peaks in dox time points (!not time 0)
dox_samples <- names(my_peaks)[!grepl("_0$", names(my_peaks))]
dox_peaks <- my_peaks[dox_samples]
dox_common_peaks <- suppressWarnings(find_common_peaks(dox_peaks))

```

Overlap of dox and non-dox (to get unique to each condition)
```{r dox vs non-dox atac peaks, dependson="previous_chunk"}
# Now overlap between dox and non-dox common peaks
dox_compare_list <- list(non_dox = non_dox_common_peaks, dox = dox_common_peaks)
dox_non_dox_ov <- suppressWarnings(find_common_peaks(dox_compare_list))

# extracting peaks unique to each condition (dox non-dox)
# Peaks unique to non_dox
unique_to_non_dox <-suppressWarnings(find_my_peaks(dox_non_dox_ov, non_dox_common_peaks))

# Peaks unique to dox
unique_to_dox <- suppressWarnings(find_my_peaks(dox_non_dox_ov, dox_common_peaks))

```

##### (d) Creating mouse gene, lincrna, mRNA annotation GRange objects

```{r creating genome annotation GRanges, dependson="previous_chunk"}

# Loading gencode genome annotation as GRanges 
gencode_gr_mouse <- rtracklayer::import("/scratch/Shares/rinnclass/MASTER_CLASS/GENOMES/M25/gencode.vM25.annotation.gtf")

# all genes
gencode_genes_mouse <- gencode_gr_mouse[gencode_gr_mouse$type == "gene"] 
gene_promoters <- promoters(gencode_genes_mouse, upstream = 2000, downstream = 2000)

# mrna_genes
mrna_genes <- gencode_genes_mouse[gencode_genes_mouse$gene_type %in% "protein_coding"]
mrna_promoters <- promoters(mrna_genes, upstream = 2000, downstream = 2000)

# lincrna_genes
lincrna_genes <- gencode_genes_mouse[gencode_genes_mouse$gene_type %in% "lincRNA"]
lincrna_gene_promoters <- promoters(lincrna_genes, upstream = 2000, downstream = 2000)

```

##### (e) Compare overlaps of dox and non-dox peaks with gene annotaitons
Now we will overlap our dox and non-dox unique peaks with genome annotations (gene promoters)
First we will find number of overlaps with gene promoters and then genes that had changed in RNAseq
```{r dox and non-dox unique peak overalp gene promoters, dependson="previous_chunk"}

# gr_list of promoters and peaks unique to non_dox condition
gr_list_gene_promoter_non_dox_ov <- list( gene_promoters = gene_promoters, non_dox_peaks = unique_to_non_dox)
non_dox_gene_promoter_ov <- suppressWarnings(find_common_peaks(gr_list_gene_promoter_non_dox_ov))

# peaks unique to dox condition overlapped with gene promoters
gr_list_gene_promoter_dox_ov <- list( gene_promoters = gene_promoters, dox_peaks = unique_to_dox)
dox_gene_promoter_ov <- suppressWarnings(find_common_peaks(gr_list_gene_promoter_dox_ov))

# Now find same gene_id in RNAseq resuts
# Loading RNAseq results from 06_Differential_expression_analyses
load("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/juke6537/MASTER_CLASS/lessons/06_Differential_expression_analyses/results/counts_results_DESEQ/DESEQ_results.rdata")

# filter significant genes from RNAseq "filtered_res_df" to non-dox unique promoter overlaps
sig_rnaseq_atac_non_dox <- non_dox_gene_promoter_ov[non_dox_gene_promoter_ov$gene_id %in% filtered_res_df$gene_id]

# filter significant genes from RNAseq "filtered_res_df" to non-dox unique promoter overlaps
sig_rnaseq_atac_dox <- non_dox_gene_promoter_ov[dox_gene_promoter_ov$gene_id %in% filtered_res_df$gene_id]

```


##### (f) use DESEQ2 to find out if peaks are changing in dox and non-dox conditions

```{r create input files and run DESEQ2 on feature counts of consensus peaks from NF_CORE, dependson="previous_chunk"}

# consensus peaks
broad_consensus_peaks <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/08_ATACseq_pipeline/00_run_pipeline/00_pipeline_run/pipeline_run/bwa/merged_library/macs2/broad_peak/consensus/consensus_peaks.mLb.clN.annotatePeaks.txt",
                             sep = "\t", header = TRUE)

# consensus peak counts
broad_consensus_counts <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/08_ATACseq_pipeline/00_run_pipeline/00_pipeline_run/pipeline_run/bwa/merged_library/macs2/broad_peak/consensus/consensus_peaks.mLb.clN.featureCounts.txt",
                             sep = "\t", header = TRUE)

# creating sample sheet "atac_samples" from file names (col names of consensus):
rownames(broad_consensus_counts) <- broad_consensus_counts$Geneid

broad_consensus_counts <- broad_consensus_counts %>%
  dplyr::select(-c(Geneid, Chr, Start, End, Strand, Length))
colnames(broad_consensus_counts) <- gsub("_REP1",
                                         "_R1",
                                         gsub(
                                           "\\.mLb\\.clN\\.sorted\\.bam",
                                           "",
                                           colnames(broad_consensus_counts)
                                         ))
count_columns <- colnames(broad_consensus_counts)

atac_samples <- data.frame(
  sample = count_columns,
  condition = ifelse(
    grepl("_0_", count_columns),
    "non-dox",
    ifelse(grepl("non-dox", count_columns), "non-dox", "dox")
  ),
  timepoint_minutes = as.numeric(sub(".*_(\\d+)_R1.*", "\\1", count_columns))
)

# Factor condition for DESEQ2 !!
atac_samples <- atac_samples %>%
  mutate(condition = factor(condition, levels = c("non-dox", "dox")))

# matrix for Deseq2
atac_dds_condition <- suppressWarnings(DESeqDataSetFromMatrix(countData = broad_consensus_counts, 
                                   colData = atac_samples, 
                                   design = ~ condition))
# Run DESeq2 condition model
atac_dds_condition <- suppressWarnings(DESeq(atac_dds_condition))

```

##### (g) extracting and analyzing DESEQ2 results
Creating a dataframe with pvlaues and lfc etc 
```{r extracting DESEQ2 results, dependson="previous_chunk"}

# Extract DESeq2 results
atac_lfc_condition <- results(atac_dds_condition) %>%
  as.data.frame() %>%
  rownames_to_column("interval_id")

# Merge with broad_consensus_peaks info

colnames(broad_consensus_peaks)[1] <- "interval_id"

atac_lfc_condition <- merge(atac_lfc_condition, 
                  broad_consensus_peaks %>%
                    dplyr::select(interval_id, Gene.Name, Nearest.PromoterID, 
                                  Distance.to.TSS, Chr, Start, End),
                  by = "interval_id")

top30_atac_genes <- atac_lfc_condition %>%
  filter(!is.na(Gene.Name)) %>%
  group_by(Gene.Name) %>%                                
  slice_max(order_by = abs(log2FoldChange), n = 1) %>%   
  ungroup() %>%
  arrange(desc(abs(log2FoldChange))) %>%                 
  slice_head(n = 30)                                    

# Print gene names and stats
top30_atac_genes %>%
  select(Gene.Name, log2FoldChange, padj, interval_id)


```

##### (h) analyzing results of DESEQ on ATACseq peak counts
Now we will see how many peaks had a padj < 0.05 and 
other features of our analyses!

```{r analysis of DESEQ2 on ATAC peak counts, dependson="previous_chunk"}

# === Your manual gene input ===
my_atac_genes_of_interest <- c("H2-Ob", "B230208B08Rik", "Setd2", "Taf7l2", "Olfr1397-ps1")

# === Filter all peaks for those genes ===
custom_atac_genes <- atac_lfc_condition %>%
  filter(Gene.Name %in% my_atac_genes_of_interest)

# === Plot with color for each interval ID ===
ggplot(custom_atac_genes, aes(x = Gene.Name, y = log2FoldChange, fill = interval_id)) +
  geom_bar(stat = "identity", color = "black", position = position_dodge(width = 0.7)) +
  geom_text(aes(label = paste0("padj = ", ifelse(is.na(padj), "NA", signif(padj, 2)))),
            position = position_dodge(width = 0.7),
            vjust = ifelse(custom_atac_genes$log2FoldChange > 0, -0.5, 1.2),
            size = 3) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  labs(
    title = "ATAC-seq log2 Fold Change of All Peaks (0â€“150 min)",
    x = "Gene", y = "log2 Fold Change", fill = "Peak (interval ID)"
  ) +
  theme_minimal()


```

